[ansible-runner] Add support to save credentials (attempt 2)

NickLaMuro opened the pull request:
Replaces https://github.com/ManageIQ/manageiq/pull/18857

Will most likely require the following being merged to implement the Machine, Network, and SCM credentials:

- https://github.com/ManageIQ/manageiq-schema/pull/382
- https://github.com/ManageIQ/manageiq/pull/18866

Effectively implements CRUD for each of the credential types, and tests that they work as expected.  Attributes that are required to be encrypted are aliased to columns where that is done already, and extra metadata or non-sensitive data is tossed into the options hash (the `host`, `domain`, `tenant`, `project`, etc. fields).


Links
-----

* 1st Pass: https://github.com/ManageIQ/manageiq/pull/18857
* Required Schema change: https://github.com/ManageIQ/manageiq-schema/pull/382
* Required Authentication class change: https://github.com/ManageIQ/manageiq/pull/18866


Steps for Testing/QA
--------------------

Similar steps to what was in #18854 but now you should be able to see the results saved in the database.

NickLaMuro commented:
@miq-bot add_labels core, enhancement, embedded_ansbile

@Fryguy @carbonin please review.

NickLaMuro commented:
bah...

@miq-bot add_label embedded ansible

carbonin commented on app/models/manageiq/providers/embedded_ansible/crud_common.rb:
```diff
@@ -55,7 +55,7 @@ def create_in_provider_queue(manager_id, params, auth_user = nil)
     private
 
     def send_notification(op_type, manager_id, params, success)
```
Since this is just a save to the local database and less of an external task I wonder if we can just ditch the notifications entirely here?

@Fryguy ?

Fryguy replied:
I'm hesitant to remove functionality that was there previously, however I made a similar comment in the previous PR, so I would be fine with ultimately removing it.

carbonin replied:
Resolving this as I think we're going to hold onto it for this pass.

NickLaMuro replied:
@carbonin For the record, we did end up ditching notifications as part of this implementation.

carbonin replied:
Ah, got it. Right, forgot that this handles the repo stuff too.

carbonin commented:
I think you should be able to handle SCM in here right? Well unless you're waiting on a decision for where to put `ssh_key_unlock`, but that isn't handled by my PRs.

NickLaMuro commented:
> Well unless you're waiting on a decision for where to put `ssh_key_unlock`, but that isn't handled by my PRs.

Yeah, that was the issue.  No place to currently put that at the moment.

carbonin commented on app/models/manageiq/providers/embedded_ansible/automation_manager/credential.rb:
```diff
@@ -12,14 +12,33 @@ class ManageIQ::Providers::EmbeddedAnsible::AutomationManager::Credential < Mana
+  end
+
+  def self.notification_excludes
+    [:password, :auth_key, :service_account]
```
Are we using service account anywhere? I thought we decided against that.

NickLaMuro replied:
These are just the 3 `encrypt_column` values found in `app/models/authentications`, and have nothing to do with what we are using in the sub models.  I just used this is a generic set of values to exclude based on the top level model that should never be included in the notifications.

Granted, this might not even matter if we are just dropping notifications for this to begin with.

Fryguy commented on app/models/manageiq/providers/embedded_ansible/automation_manager/credential.rb:
```diff
@@ -12,14 +12,33 @@ class ManageIQ::Providers::EmbeddedAnsible::AutomationManager::Credential < Mana
+  end
+
+  def self.notification_excludes
+    [:password, :auth_key, :service_account]
```
This list is already enumerated elsewhere via `encrypted_columns` class method...it's probably preferable to just put that list right in the base class, and then you don't need these overrides at all

NickLaMuro replied:
@Fryguy I think what you are suggesting here and pushing this up to `CrudCommon` will work, but I don't think this will work in the specialized classes, since a lot of them make use of `alias_attribute`, which `PasswordMixin` does not support.

Fryguy commented on app/models/manageiq/providers/embedded_ansible/crud_common.rb:
```diff
@@ -66,6 +66,11 @@ def send_notification(op_type, manager_id, params, success)
+
+    # Override in subclass if necessary
+    def notification_excludes
+      [:name, :manager_ref]
```
As per my comment above, here we can just do `[:name, :manager_ref] + Authentication.encrypted_columns`, which I think covers everythings.

NickLaMuro replied:
See my response to your other comment.

Fryguy commented on app/models/manageiq/providers/embedded_ansible/automation_manager/azure_credential.rb:
```diff
@@ -47,7 +47,36 @@ class ManageIQ::Providers::EmbeddedAnsible::AutomationManager::AzureCredential <
     :attributes => API_ATTRIBUTES
   }.freeze
 
+  alias_attribute :secret, :auth_key
```
Why is this needed?  I'm concerned because it exposes an encrypted column via an alternate attribute, which does things like expose it to reporting, for example.  There are a number of code paths that expect that `Model.encrypted_columns` holds the list.

NickLaMuro replied:
`:secret` is what is passed into from the UI, so we need to support it in some way.  Possibly this isn't required if we just are doing `params_to_attributes` however.  I can look into that.

Fryguy replied:
We can also change the UI to not send secret, but instead send auth_key as we expect

NickLaMuro commented:
@Fryguy @carbonin So... after much rebasing and probably giving myself more work than it is worth, I come back to you with.... options!

Behind[ _**"Door #1"**_](https://github.com/ManageIQ/manageiq/pull/18870), we have this PR, now with extra credentials added (aka:  no real changes).

In [_**"Door #2"**_](https://github.com/ManageIQ/manageiq/compare/master...NickLaMuro:ansible-runner-initial-credential-save-v3), we have this PR, but without notifications and `alias_attribute`!!

And in [_**"Door #3"**_](https://github.com/ManageIQ/manageiq/compare/master...NickLaMuro:ansible-runner-initial-credential-save-without-queue), we take things a step further and remove the queue all together!


* * *

Okay, kidding and unnecessary ceremony aside, I just decided to put together a few permutations of this PR up for review to show what they might look like, and am curious on your thoughts on a direction for it.  The last one is the most aggressive my completely removing the notification and queue elements that were there previously, with the middle branch sort of somewhere in the middle.

In regards to [@Fryguy's comment](https://github.com/ManageIQ/manageiq/pull/18870#discussion_r293960970):

> We can also change the UI to not send secret, but instead send auth_key as we expect

I didn't go that route in any of the options since I am trying to at least keep elements that QE are expecting the same for this first pass.  For what it is worth, I don't think it would take much effort and would just be changing `API_OPTIONS`, so if we are going full "breaking change" with this then it won't be a hard change (however, I have been on the bad side of those lately with this effort, so not terribly keen on changing the UI just for the heck of it).

Anyway, let me know which you think is a good path to go with, and we can move forward with that...

In the mean time, I have to go fix something else that is busted with the `ansible-runner` stuff... 😩

NickLaMuro commented:
Worth noting:  Based on the bug that I fixed here:

https://github.com/ManageIQ/manageiq/pull/18897

We probably want to make sure we are also adding a `manager_ref` to all credentials that are created, so I might circle back and make sure we do that in whatever solution we go decide on.

Fryguy commented:
I vote Door 2

NickLaMuro commented:
Door #2 two it is!

I have updated this branch to use those changes, and archived the previous changes here:

https://github.com/NickLaMuro/manageiq/tree/ansible-runner-initial-credential-save-v2_old

And removed the `[WIP]` from this PR.

carbonin commented on app/models/manageiq/providers/embedded_ansible/automation_manager/azure_credential.rb:
```diff
@@ -47,7 +47,33 @@ class ManageIQ::Providers::EmbeddedAnsible::AutomationManager::AzureCredential <
   end
+
+  def self.params_to_attributes(params)
+    attrs = params.reverse_merge(:options => {})
```
I'm not sure what behavior we actually want here. Do we want options to be empty if none are provided or `nil`?

When migrating, I set it to `nil` if there are no options for a credential, even if we *could* have options for that type.

NickLaMuro replied:
I guess the problem with that is then we have to do extra existence checks when setting `params[:options][:my_val]` for the passed in `params[:options]`, or conditionally set `attrs[:options] = {}` if any of the expected keys exist.

Also, the helper methods (like `#client`, `#tenant`, and `#subscription`) become more complex, as we then need to do `nil` checks for each.

* * *

There are cons I am sure for either implementation, and I am not hard set on using what I have here, but that is the rational I used when putting what I have here together.

Let me know if you think we should change it.

carbonin replied:
Well the column is nullable and we don't set a default value on the model so I don't think we can assume that it will always have a value regardless of what we do here.

NickLaMuro replied:
Fixed.
